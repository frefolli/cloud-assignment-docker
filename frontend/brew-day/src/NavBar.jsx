function NavBar() {
    return(
        <ul>
            <li><a href="#"><img src="../icons/logo1.png" height="50px"></img></a></li>
            <li><a href="#">Ricette</a></li>
            <li><a href="#">Birre</a></li>
            <li><a href="#">Inventario</a></li>
            <li><a href="#">Spesa</a></li>
        </ul>
    );
}

export default NavBar;