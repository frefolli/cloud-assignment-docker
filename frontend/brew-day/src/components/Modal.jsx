import React from 'react';

const Modal = ({ children, showModal, setShowModal }) => {
    const handleModalClick = (e) => {
        e.stopPropagation();
    };

    return (
        <div>
            {showModal ? (
                <div className={`modal show`} onClick={() => setShowModal(false)} data-testid="modal" style={{overflowY: "auto",}}>
                    <div className="modal-content" onClick={handleModalClick} style={{width: "75%"}}>
                        {children}
                    </div>
                </div>
            ) : (
                <div className={`modal hide`} onClick={() => setShowModal(false)} data-testid="modal"></div>
            )}
        </div>
    );
}

export default Modal;