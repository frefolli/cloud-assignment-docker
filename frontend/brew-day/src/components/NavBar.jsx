import React from "react";
import {BrowserRouter as Router, Routes, Route, Link} from "react-router-dom";
import Birre from "../pages/Birre";
import Home from "../pages/Home";
import Inventario from "../pages/Inventario";
import Ricette from "../pages/Ricette";
import Spesa from "../pages/Spesa";
import Impostazioni from "../pages/Impostazioni";

const NavBar = () => {
    return (
        <Router>
            <div>
                <ul id="navBar">
                    <li>
                        <Link to="/">
                            <img alt="navlogo" id="logo-image" src="../icons/logo2.png"/>
                            <span id="logo-text">Brew Day!</span>
                        </Link>
                    </li>
                    <li>
                        <Link to="/ricette">Ricette</Link>
                    </li>
                    <li>
                        <Link to="/birre">Birre</Link>
                    </li>
                    <li>
                        <Link to="/inventario">Inventario</Link>
                    </li>
                    <li>
                        <Link to="/spesa">Spesa</Link>
                    </li>
                    <li>
                        <Link to="/impostazioni">Impostazioni</Link>
                    </li>
                </ul>
            </div>

            <Routes>
                <Route path="/" element={<Home />}/>
                <Route path="/ricette" element={<Ricette />}/>
                <Route path="/birre" element={<Birre />}/>
                <Route path="/inventario" element={<Inventario />}/>
                <Route path="/spesa" element={<Spesa />}/>                
                <Route path="/impostazioni" element={<Impostazioni />}/>
            </Routes>
        </Router>
    );
}

export default NavBar;