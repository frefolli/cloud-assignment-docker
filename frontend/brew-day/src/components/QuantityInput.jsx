import React from "react";

class QuantityInput extends React.Component {
  constructor() {
    super();
    this.state = { value: '' };
    this.handleValue = this.handleValue.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.comma = null;
    this.max = 999;
  }
  handleValue(event) {
    const string = event.target.value;

    if (string == '.') this.setState({ value: '0.'});
    else if (string.endsWith('.') && !this.comma) this.setState({ value: string });
    else if (string.startsWith('0') && string.endsWith('00') && !string.includes('.')) this.setState({ value: '0' });
    else if (/[0-9]$/.test(string)) {
        if(parseFloat(string) <= 999) this.setState({ value: string });
        else this.setState({ value: this.max });
    }
    else this.setState({ value: string.slice(0, -1) });

    this.comma = string.includes('.');
  }
  handleBlur(event) {
    if(/0\.0+$/.test(event.target.value)) this.setState({ value: '0'});
  }
  render() {
    return (
      <div>
        <input
          id="shoppingItemId"
          name="shoppingItemId"
          value={this.state.value}
          onChange={this.handleValue}
          onBlur={this.handleBlur}
        ></input>
      </div>
    );
  }
}

export default QuantityInput;