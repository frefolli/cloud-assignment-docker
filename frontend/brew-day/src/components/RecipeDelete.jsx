import React, { Component }  from "react";
import { ThemeProvider } from "@mui/material";
import theme from "../theme/theme";
import MButton from '../components/MButton';

class RecipeDelete extends Component{
  constructor(props) {
    super(props);
    this.state = {name: "", description: "", ingredients: []};
    this.deleteRecipe = this.deleteRecipe.bind(this);
      this.triggerReload = this.triggerReload.bind(this);
  }

  triggerReload() {
      fetch(`/api/recipes/${this.props.recipeID}`)
      .then(response => response.json())
      .then(data => this.setState({...data}));
  }

  componentDidMount() {
    this.triggerReload();
  }

  render(){
    return (
      <ThemeProvider theme={theme}>
        <div>
          <center>
            <h1>{this.state.name}</h1>
            <p>{this.state.description}</p>
            <p>Sei sicuro di voler rimuovere la ricetta?</p>
            <MButton text="Conferma" onClick={() => this.deleteRecipe(this.state.recipeID)} />
          </center>
        </div>
      </ThemeProvider>
    );
  }

  deleteRecipe(id) {
    fetch(`/api/recipes/${id}`, {
        method: 'DELETE',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    });
    this.props.onConfirm();
}

}

export default RecipeDelete;
