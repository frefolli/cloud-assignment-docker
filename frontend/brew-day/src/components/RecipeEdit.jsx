import React, { Component } from "react";
import { ThemeProvider } from "@mui/material";
import theme from "../theme/theme";
import MButton from '../components/MButton';

class RecipeEdit extends Component{

  constructor(props) {
    super(props);
    this.state = {newIngredientName: null, newIngredientQuantity: null, ...props};
  }

  setQuantity(id, event){
    let quantity = event.target.value;
    this.setState({ingredients:     
      this.state.ingredients.map( item => {
        if(item.ingredientID === id){
          item.quantity=quantity;
        }
        return item;
      })
    })
  }

  setName(event){
    let newNameRecipe = event.target.value;
    this.setState({name: newNameRecipe})
  }

  setDescription(event){
    let newDescription = event.target.value;
    this.setState({description: newDescription})
  }

  setNewIngredientName(event){
    let newName = event.target.value;
    this.setState({newIngredientName: newName});
  }

  setNewIngredientQuantity(event){
    let newQuantity = event.target.value;
    this.setState({newIngredientQuantity: newQuantity});
  }

  render(){
    const itemList = this.state.ingredients.map(item => {
      return <tr key={item.ingredientID}>
          <td>{item.name}</td>
          <td><input value={item.quantity} type="text" style={{width: "50%", textAlign:"center"}} onChange={ (event) => this.setQuantity(item.ingredientID, event)}></input></td>
          <td>
            <MButton text="V" onClick={() => this.editQuantity(item.ingredientID)} />
            <MButton text="X" onClick={() => this.deleteIngredient(item.ingredientID)} />
          </td>
        </tr>
    });
  
    return (
      <ThemeProvider theme={theme}>
        <div>
          <center>
          <table className="myTable">
              <tbody> 
                <tr>
                  <td> <p>Nome Ricetta:</p> </td>
                  <td><input id="nameTextArea" value={this.state.name} style={{width: "80%"}} onChange={ (event) => this.setName(event)}></input></td>
                  <td>
                    <MButton text="V" onClick={() => this.editName()} />
                  </td>        
                </tr>
              </tbody>
            </table>
            <table className="myTable">
              <tbody> 
                <tr>
                    <td><p>Descrizione:</p></td>
                    <td><textarea id="descriptionTextArea" value={this.state.description} style={{width: "80%"}} onChange={ (event) => this.setDescription(event)}></textarea></td>
                    <td>
                      <MButton text="V" onClick={() => this.editDescription()} />
                    </td>
                </tr>
              </tbody>
            </table>
            
            <table className="myTable">
              <thead>
                <tr>
                  <th width="30%">Nome</th>
                  <th width="30%">Quantit√†</th>
                  <th width="30%">Azioni</th>
                </tr>
              </thead>
              <tbody>
                {itemList}
                <tr>
                  <td><input value={null} type="text" style={{width: "50%", textAlign:"center"}} onChange={ (event) => this.setNewIngredientName(event)}></input></td>
                  <td><input value={null} type="text" style={{width: "50%", textAlign:"center"}} onChange={ (event) => this.setNewIngredientQuantity(event)}></input></td>
                  <MButton text="V" onClick={() => this.addIngredient()} />
                </tr>
              </tbody>
            </table>
          </center>
        </div>
      </ThemeProvider>
    );
  }

  deleteIngredient(id) {
    fetch(`/api/recipes/${this.state.recipeID}/${id}`, {
        method: 'DELETE',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    });
  }

  editQuantity(id) {
    let newQuantity = [...this.state.ingredients].filter(i => i.ingredientID === id)[0].quantity;
    fetch(`/api/recipes/${this.state.recipeID}/${id}`, {
        method: 'PUT',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({quantity: newQuantity})
    });
  }

  editName() {
    fetch(`/api/recipes/${this.state.recipeID}`, {
        method: 'PUT',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({name: this.state.name})
    });
  }

  editDescription() {
    fetch(`/api/recipes/${this.state.recipeID}`, {
        method: 'PUT',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({description: this.state.description})
    });
  }

  addIngredient() {
    fetch(`/api/recipes/${this.state.recipeID}`, {
      method: 'POST',
      headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json'
      },
      body: JSON.stringify({name: this.state.newIngredientName, quantity: this.state.newIngredientQuantity})
  });
  }

}


export default RecipeEdit;