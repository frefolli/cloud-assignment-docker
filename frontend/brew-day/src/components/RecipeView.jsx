import React from "react";

const RecipeView = (props) => {

  const itemList = props.ingredients.map(item => {
    let imagePath = `../../icons/inventory-icons/${item.name}.png`;
    let defaultImage = "../../icons/inventory-icons/sconosciuto.png";
    return <tr key={item.name}>
        <td>
          <img
            className="shoppingImage"
            src={imagePath}                 
            onError={(e) => { e.target.onerror = null; e.target.src=defaultImage }}
          />
        </td>
        <td>{item.name}</td>
        <td>{item.quantity}</td>
      </tr>
  });

  return (
    <div>
      <center>
        <h1> {props.name}</h1>
        <p>In RecipeView devo cambiare questo testo con props.description tra graffe per avere la descrizione</p>
        <table className="myTable">
          <thead>
            <tr>
              <th width="25%">Immagine</th>
              <th width="25%">Nome</th>
              <th width="25%">Quantit√†</th>
            </tr>
          </thead>
          <tbody>
            {itemList}
          </tbody>
        </table>
      </center>
    </div>
  );
};

export default RecipeView;