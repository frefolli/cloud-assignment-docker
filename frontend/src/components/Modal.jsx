import * as React from 'react';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import Paper from '@mui/material/Paper';
import Draggable from 'react-draggable';
import MButton from './MButton';

function PaperComponent(props) {
  return (
    <Draggable
      handle="#draggable-dialog-title"
      cancel={'[class*="MuiDialogContent-root"]'}
    >
      <Paper {...props} />
    </Draggable>
  );
}

/**
 * This component represents a modal dialog for displaying content.
 * 
 * @class Modal
 * 
 * @param {Object} props - The component's properties.
 * @param {boolean} props.showModal - If `true`, the modal is displayed; otherwise, it's hidden.
 * @param {function} props.setShowModal - A function to control the visibility of the modal.
 * @param {JSX.Element} props.children - The content to be displayed within the modal.
 * 
 * @return {JSX.Element} A modal dialog for displaying content.
 * 
 * @example
 * // Example usage of the Modal component
 * <Modal
 *   showModal={isModalVisible}
 *   setShowModal={setModalVisibility}
 * >
 *   <div>
 *     <h2>Modal Content</h2>
 *     <p>This is the content of the modal dialog.</p>
 *   </div>
 * </Modal>
 */

export default class Modal extends React.Component {
  render() {
    return (
        <Dialog
          open={this.props.showModal}
          onClose={() => this.props.setShowModal(false)}
          PaperComponent={PaperComponent}
          aria-labelledby="draggable-dialog-title"
          fullWidth={true}
          maxWidth={'md'}
          className='modale'
          aria-label='Modal'
        >
          <DialogTitle style={{ cursor: 'move' }} id="draggable-dialog-title">
          </DialogTitle>
          <DialogContent>
          {this.props.children}
          </DialogContent>
          <DialogActions>
            <MButton autoFocus onClick={() => this.props.setShowModal(false)} text="Cancel"/>
          </DialogActions>
        </Dialog>
    );
  }
}
