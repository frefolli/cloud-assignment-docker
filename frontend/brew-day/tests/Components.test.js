import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent } from "@testing-library/react";
import NavBar from "../src/components/NavBar";
import QuantityInput from "../src/components/QuantityInput";
import Modal from "../src/components/Modal";
import RecipeView from "../src/components/RecipeView";
import RecipeEdit from "../src/components/RecipeEdit";
import RecipeDelete from "../src/components/RecipeDelete";

const testIngredient = [{ name: "name", quantity: "0" }];

describe("NavBar component", () => {
  test("should render correctly", () => {
    const { container } = render(<NavBar />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("QuantityInput component", () => {
  test("should render correctly", () => {
    const { container } = render(<QuantityInput />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("correctly sets the value of the input element", () => {
    const { container } = render(<QuantityInput />);
    const input = container.querySelector("input");

    fireEvent.change(input, { target: { value: "100" } });
    expect(input.value).toBe("100");

    fireEvent.change(input, { target: { value: "1000" } });
    expect(input.value).toBe("999");

    fireEvent.change(input, { target: { value: "100.5" } });
    expect(input.value).toBe("100.5");
  });

  test("correctly sets the value of the input element on blur", () => {
    const { container } = render(<QuantityInput />);
    const input = container.querySelector("input");

    fireEvent.change(input, { target: { value: "0.00" } });
    fireEvent.blur(input);
    expect(input.value).toBe("0");
  });

  test("Input only accepts numbers", () => {
    const component = render(<QuantityInput />);
    fireEvent.change(component.getByTestId("quantity-input"), {
      target: { value: "a" },
    });
    expect(component.getByTestId("quantity-input").value).toBe("");
  });

  test("Input does not allow minus sign, so negative numbers", () => {
    const component = render(<QuantityInput />);
    fireEvent.change(component.getByTestId("quantity-input"), {
      target: { value: "-" },
    });
    expect(component.getByTestId("quantity-input").value).toBe("");
  });

  test("If input starts with . it automatically returns 0.", () => {
    const component = render(<QuantityInput />);
    fireEvent.change(component.getByTestId("quantity-input"), {
      target: { value: "." },
    });
    expect(component.getByTestId("quantity-input").value).toBe("0.");
  });

  test("999 is the max input", () => {
    const component = render(<QuantityInput />);
    fireEvent.change(component.getByTestId("quantity-input"), {
      target: { value: "1000" },
    });
    expect(component.getByTestId("quantity-input").value).toBe("999");
  });

  test("00 is rendered to 0", () => {
    const component = render(<QuantityInput />);
    fireEvent.change(component.getByTestId("quantity-input"), {
      target: { value: "00" },
    });
    expect(component.getByTestId("quantity-input").value).toBe("0");
  });

  test("If the first comma comes after a digit or more, keep the number", () => {
    const component = render(<QuantityInput />);
    fireEvent.change(component.getByTestId("quantity-input"), {
      target: { value: "0." },
    });
    expect(component.getByTestId("quantity-input").value).toBe("0.");
  });

  test("0.0 renders to 0", () => {
    const component = render(<QuantityInput />);
    fireEvent.change(component.getByTestId("quantity-input"), {
      target: { value: "0.0" },
    });
    fireEvent.blur(component.getByTestId("quantity-input"));
    expect(component.getByTestId("quantity-input").value).toBe("0");
  });
});

describe("RecipeDelete component", () => {
  test("should render correctly", () => {
    const { container } = render(
      <RecipeDelete
        name="name"
        description="description"
        ingredients={testIngredient}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("RecipeEdit component", () => {
  test("should render correctly", () => {
    const { container } = render(
      <RecipeEdit
        name="name"
        description="description"
        ingredients={testIngredient}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("RecipeView component", () => {
  test("should render correctly", () => {
    const { container } = render(
      <RecipeView
        name="name"
        description="description"
        ingredients={testIngredient}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Modal component", () => {
  test("should render correctly", () => {
    const { container } = render(<Modal />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("handleModalClick should stopPropagation", () => {
    const { getByTestId } = render(
      <Modal showModal={true} setShowModal={() => {}}>
        <div data-testid="modal-content"></div>
      </Modal>
    );
    const modalContent = getByTestId("modal-content");

    fireEvent.click(modalContent);

    expect(modalContent).toBeInTheDocument();
  });

  test("clicking on the modal sets showModal to false", () => {
    const setShowModal = jest.fn();
    const { getByTestId } = render(
      <Modal showModal={true} setShowModal={setShowModal}>
        <div data-testid="modal-content"></div>
      </Modal>
    );
    fireEvent.click(getByTestId("modal"));
    expect(setShowModal).toHaveBeenCalledWith(false);
  });

  test("clicking on the modal sets showModal to false", () => {
    const setShowModal = jest.fn();
    const { getByTestId } = render(
      <Modal showModal={false} setShowModal={setShowModal}>
        <div data-testid="modal-content"></div>
      </Modal>
    );
    fireEvent.click(getByTestId("modal"));
    expect(setShowModal).toHaveBeenCalledWith(false);
  });
});
