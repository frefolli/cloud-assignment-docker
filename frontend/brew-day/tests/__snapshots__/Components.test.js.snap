// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
<span
  class="shoppingItemText"
>
  0
</span>
`;

exports[`Modal component should render correctly 1`] = `
<div>
  <div
    class="modal hide"
    data-testid="modal"
  />
</div>
`;

exports[`NavBar component should render correctly 1`] = `
<div>
  <ul
    id="navBar"
  >
    <li>
      <a
        href="/"
      >
        <img
          id="logo-image"
          src="../icons/logo2.png"
        />
        <span
          id="logo-text"
        >
          Brew Day!
        </span>
      </a>
    </li>
    <li>
      <a
        href="/ricette"
      >
        Ricette
      </a>
    </li>
    <li>
      <a
        href="/birre"
      >
        Birre
      </a>
    </li>
    <li>
      <a
        href="/inventario"
      >
        Inventario
      </a>
    </li>
    <li>
      <a
        href="/spesa"
      >
        Spesa
      </a>
    </li>
  </ul>
</div>
`;

exports[`QuantityInput component should render correctly 1`] = `
<div>
  <input
    data-testid="quantity-input"
    id="shoppingItemId"
    name="shoppingItemId"
    value=""
  />
</div>
`;

exports[`RecipeDelete component should render correctly 1`] = `
<div>
  <div>
    <h1>
      name
    </h1>
    <p>
      description
    </p>
    <table
      class="recipeView"
    >
      <thead>
        <tr>
          <th>
            Nome
          </th>
          <th>
            Quantità
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            name
          </td>
          <td>
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    class="recipeButton"
  >
    Sei sicuro di voler rimuovere la ricetta?
  </button>
</div>
`;

exports[`RecipeEdit component should render correctly 1`] = `
<div>
  <div>
    <div
      id="recipeEditDiv"
    >
      <p>
        Nome:
      </p>
      <textarea
        id="nameTextArea"
      >
        name
      </textarea>
      <p>
        Descrizione:
      </p>
      <textarea
        id="descriptionTextArea"
      >
        description
      </textarea>
      <br />
      <table
        class="recipeView"
      >
        <thead>
          <tr>
            <th>
              Ingrediente
            </th>
            <th>
              Quantità
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <p>
                Acqua
              </p>
            </td>
            <td>
              <textarea />
            </td>
          </tr>
          <tr>
            <td>
              <p>
                Additivi
              </p>
            </td>
            <td>
              <textarea />
            </td>
          </tr>
          <tr>
            <td>
              <p>
                Lievito
              </p>
            </td>
            <td>
              <textarea />
            </td>
          </tr>
          <tr>
            <td>
              <p>
                Luppoli
              </p>
            </td>
            <td>
              <textarea />
            </td>
          </tr>
          <tr>
            <td>
              <p>
                Malto
              </p>
            </td>
            <td>
              <textarea />
            </td>
          </tr>
          <tr>
            <td>
              <p>
                Zuccheri
              </p>
            </td>
            <td>
              <textarea />
            </td>
          </tr>
        </tbody>
      </table>
      <button>
        Modifica ricetta
      </button>
    </div>
  </div>
</div>
`;

exports[`RecipeView component should render correctly 1`] = `
<div>
  <h1>
    name
  </h1>
  <p>
    description
  </p>
  <table
    class="recipeView"
  >
    <thead>
      <tr>
        <th>
          Nome
        </th>
        <th>
          Quantità
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          name
        </td>
        <td>
          0
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
